scalding-core/src/main/scala/com/twitter/scalding/serialization/JavaStreamEnrichments.scala

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2015 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package com.twitter.scalding.serialization
+
+import java.io.{ ByteArrayInputStream, ByteArrayOutputStream, InputStream, OutputStream, EOFException }
+
+object JavaStreamEnrichments {
+  def eof: Nothing = throw new EOFException()
+
+  implicit class RichByteArrayOutputStream(val baos: ByteArrayOutputStream) extends AnyVal {
+    def toInputStream: ByteArrayInputStream = new ByteArrayInputStream(baos.toByteArray)
```

Comment:
baos.toByteArray involves a copy of the internal buffer used by ByteArrayOutputStream -- if the usage pattern is to write to a ByteArrayOutputStream, then fully consume it as an input stream, then we could avoid the copy by creating a ByteArrayInputOutputStream or something like that.

Also, since we care a lot about performance here, be aware that ByteArrayOutputStream works like Vector in that it "grows" by allocating a new buffer twice as big as the current then does a full copy. So choosing a good starting size is important, or for writing large amounts of data, using an alternative strategy like allocating byte array slabs into a an ArrayList (no copying) might make sense.